### PR TITLE
Test cases for LAG on Distributed VOQ System

### PR DESCRIPTION
**Pull request by oxygen980**
### Description of PR
Test cases for LAG 
Test plan PR :-
https://github.com/Azure/sonic-mgmt/pull/3033

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x ] Test case(new/improvement)

#### What is the motivation for this PR?
Test cases for LAG on Distributed VOQ System

#### How did you verify/test it?
Verified on T2 topology 
#### Supported testbed topology if it's a new test case?
T2 topology

### CODE REVIEW

**Comment by oxygen980** [comment]
retest vs image please

**Comment by oxygen980** [comment]
@ygrik The run is failing in 'test_po_update' duting LogAnalyzer. My code changes are not touching this file. Could you please re-run the build ?

**Comment by shubav** [comment]
@yxieca, Could you help review these test cases? The failures in the test case are unrelated to this commit. Could you help restart the pipeline?

**Comment by oxygen980** [comment]
@yxieca, Could you help restart the pipeline? the failure is not from the code change.

**Comment by shubav** [comment]
The pipeline checks have passed.
@yxieca, @wangxin - could you please review?
@eswaranb - could you please add Arista reviewers?
There are just 4 TCs. Thanks.

**Comment by anshuv-mfst** [comment]
@yxieca @wangxin , @saravanansv - could you please take a look, thanks.

File: tests/common/helpers/voq_lag.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    """
+    appdb = AppDbCli(asic)
+    app_db_lag_list = appdb.get_app_db_lag_list()
+    if deleted:
+        for lag in app_db_lag_list:
+            if TMP_PC not in lag:
+                logging.info('LAG {} is deleted in ASIC app db'.format(TMP_PC))
+                return
+        pytest.fail('LAG {} still exist in ASIC app db,'
+                    ' Expected was should be deleted from asic app db.'.format(TMP_PC))
```
Comment (wangxin) [inline tests/common/helpers/voq_lag.py:129]: Is there a logic issue here?
If `deleted` is `True`, then the current code would return if `TMP_PC` not in the first lag in `app_db_lag_list`. However, it is possible that `TMP_PC` in rest of the lag in `app_db_lag_list`. So, the correct logic should be like below, right?
```
    if deleted:
        for lag in app_db_lag_list:
            if TMP_PC in lag:
                pytest.fail('LAG {} still exist in ASIC app db,'
                            ' Expected was should be deleted from asic app db.'.format(TMP_PC))
        logging.info('LAG {} is deleted in ASIC app db'.format(TMP_PC))
        return

```

File: tests/common/helpers/voq_lag.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+            duthost.shell("config portchannel {} member add {} {}"
+                          .format(asic.cli_ns_option, portchannel, member))
+
+    if portchannel_ip:
+        duthost.shell("config interface {} ip add {} {}"
+                      .format(asic.cli_ns_option, portchannel, portchannel_ip))
+        int_facts = duthost.interface_facts(namespace=asic.namespace)['ansible_facts']
+        pytest_assert(int_facts['ansible_interface_facts']
+                      [portchannel]['ipv4']['address'] == portchannel_ip.split('/')[0])
+        time.sleep(30)
```
Comment (wangxin) [inline tests/common/helpers/voq_lag.py:86]: There are multiple `time.sleep(30)` in this script. Can you replace the hard sleep with utility like `wait_until`? Then we may be able to steal some seconds of running the test script.

File: tests/common/helpers/voq_lag.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                            'Expected was should be deleted from asic app db.'.format(TMP_PC))
+
+        logging.info('For lag {} lag members {} are deleted in ASIC app db'.format(TMP_PC, pc_members))
+    else:
+        for member in pc_members:
+            pattern = "{}:{}".format(TMP_PC, member)
+            exist = False
+            for lag in app_db_lag_member_list:
+                if pattern in lag:
+                    exist = True
```
Comment (wangxin) [inline tests/common/helpers/voq_lag.py:247]: Add `break` to jump out from the iteration early?

File: tests/common/helpers/voq_lag.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    """
+    appdb = AppDbCli(asic)
+    app_db_lag_member_list = appdb.get_app_db_lag_member_list()
+    if deleted:
+        for member in pc_members:
+            pattern = "{}:{}".format(TMP_PC, member)
+            exist = False
+            for lag_member in app_db_lag_member_list:
+                if pattern in lag_member:
+                    exist = True
```
Comment (wangxin) [inline tests/common/helpers/voq_lag.py:234]: Add `break` to jump out from the iteration early?

File: tests/common/helpers/voq_lag.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+    for sup in duthosts.supervisor_nodes:
+        voqdb = VoqDbCli(sup)
+        lag_member_list = voqdb.get_lag_member_list()
+        if deleted:
+            for member in members:
+                exist = False
+                pattern = "{}.*{}".format(TMP_PC, member)
+                for lag_member in lag_member_list:
+                    if re.search(pattern, lag_member):
+                        exist = True
```
Comment (wangxin) [inline tests/common/helpers/voq_lag.py:321]: Add `break` to jump out from the iteration early?

File: tests/common/helpers/voq_lag.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+        else:
+            for member in members:
+                exist = False
+                pattern = "{}.*{}".format(TMP_PC, member)
+                for lag_member in lag_member_list:
+                    if re.search(pattern, lag_member):
+                        exist = True
+                        logging.info('lag member {} found in system lag member table {}'
+                                     .format(member, lag_member))
```
Comment (wangxin) [inline tests/common/helpers/voq_lag.py:337]: Add `break` to jump out from the iteration early?

**Review by wangxin — APPROVED** [review]

**Comment by skbarista** [comment]
Hi, This is Sambath from Arista. Sorry I am little late on this. The test plan from https://github.com/Azure/sonic-mgmt/pull/3033/files as well as test cases does not seem to have any traffic validation. Would traffic validation be added in a future update to the test plan/test cases?

**Comment by oxygen980** [comment]
@skbarista Thank you for the review ! As we have mentioned in the test plan 'This test plan does not cover functioning of LAG as an IP/Layer3 endpoint as that is outside the scope of the code PR' we are not going to have traffic test cases in future too.

**Comment by shubav** [comment]
> @skbarista Thank you for the review ! As we have mentioned in the test plan 'This test plan does not cover functioning of LAG as an IP/Layer3 endpoint as that is outside the scope of the code PR' we are not going to have traffic test cases in future too.

Also, this test case will execute on a T2 topology, of which, LAG is a part. Many T1 test cases, when converted by the chassis subgroup, will cover this forwarding aspect.

We kept forwarding out of this because:
1. The code PR does not touch fwding and not even teamd
2.  Other forwarding tests check hashing and forwarding. So we avoided duplication

**Comment by skbarista** [comment]
Thanks for the clarification @shubav @oxygen980

**Comment by shubav** [comment]
@wangxin comments from you and Arista have been addressed. If it looks fine, could you help get this merged, please?

**Pull request merged**